docs/DesignDocs/DebuggerSupportForMultiModuleEnclaves.md

```diff
@@ -0,0 +1,268 @@
+		uint64_t path_length;
+
+		// Memory region where the enclave has been loaded.
+		uint64_t base_address;
+		uint64_t size;
+	} oe_debug_module_t;
+	
+```
+
+The `path` and `path_length` together specify the path of the module's ELF
+binary. The `base_address` and `size` together specify the memory region
+where the enclave has been loaded.
+
+The following notification functions are used to notify the debugger about the
+loading and unloading of a module.
+
+- `oe_result_t oe_result_t oe_debug_notify_module_loaded(oe_debug_module_t* module)`
+
+   This function is called by the SDK's host-runtime after a module has been
+   loaded, but before it has been initialized. A module is initialized by invoking
+   it's constructor functions. It doesn not matter whether the module has been relocated
+   or not before the debugger is notified about the module.
+
+   In situations where a module may be dynamically loaded by an enclave, as is the case
+   when SGX LKL/openlibos launch applications within the enclave, this notification
+   function is called by the SDK's enclave-side runtime prior to the initialization
+   of the module.
+
+- `oe_result_t oe_debug_notify_module_unloaded(oe_debug_module_t* module)`
+
+   This function is called by the SDK runtimes (host-side or enclave-side), after a
+   module has been unloaded.
+
+Debugger Handling of Module Load/Unload Notifications
+
+- Linux
+  Similar to enclave load/unload, the debugger sets internal breakpoints in
+  `oe_debug_notify_module_loaded` and `oe_debug_notify_module_unloaded`
+  functions. When the SDK runtimes call these functions, the breakpoint is
+  triggered and the debugger get control. The debugger reads the module information
+  from the `oe_debug_module_t` structure either loads or unloads the debugging
+  information for the module.
+
+- Windows
+  - Host Side
+  
+    The SDK host-side runtime uses the `RaiseException` Windows API to
+	communicate with the debugger.
+	`oe_debug_notify_module_loaded` raises the following exception:
+	```c
+ 
+        ULONG_PTR args[1] = {(ULONG_PTR)module};
+		RaiseException(
+		  OE_DEBUGRT_MODULE_LOADED_EVENT,
+		  0, // dwFlags
+		  1, // number of args
+		  args);
+		  
+    ```
+	
+     whereas `oe_debug_notify_enclave_terminated` raises:
+     ```c
+   
+        ULONG_PTR args[1] = {(ULONG_PTR)module};
+        RaiseException(
+          OE_DEBUGRT_MODULE_UNLOADED_EVENT,
+          0, // dwFlags
+          1, // number of args
+          args);
+		  
+	```
+	
+  - Enclave side
+    
+	The enclave side runtime cannot call Windows API functions. Instead, it
+	relies on `int 3` exception to communicate with the debugger. The `int 3`
+	instruction is preceeded by special byte sequence `0x0edeb0gca11bac00` to
+	allow the debugger to distinguish between these special `int 3` exceptions
+	and any other `int 3` exceptions that may occur.
+	
+	`oe_debug_notify_module_loaded` uses the following instruction pattern:
+	```asm
+
+       mov module, %rax
```

Comment:
The same ELF file should run on both Windows and Linux, so you have to insert this at compile time without know whether it will run on Windows or Linux or both.

Reply:
In the windows case we should be able to rely on symbols for the enclave portion (since they should be embedded in the ELF file). That way both the Linux and Windows in-enclave load can be handled the same way via internal breakpoints.